### PR TITLE
fix(inductor): match DXP usable LX reservation

### LINKED ISSUE #3222: Correction issue due to backend LX buffer overflow
### What happened?

Compiled ops whose per-core LX scratchpad usage reaches **1600KB** (97.7% of the 1638.4KB / 80%-of-2MB frontend partition) produce silently corrupted output on device — no error or warning anywhere in the compile pipeline.

Through env var `DXP_LX_FRAC_AVAIL`, 0.2 by default, the frontend (torch-spyre) and backend (`dxp_standalone`) agree on a fixed partition of the 2MB per-core LX scratchpad: `[0, 1638.4KB)` for the frontend, `[1638.4KB, 2048KB)` for the backend. We observed that (with `DXP_VERBOSE=1`) the backend's own finalized schedule tree shows it placing an internal LX allocation at address **`1625344` bytes (1587.25KB)** - **overflown from backend's partition**. When a frontend buffer is still live at that address (which happens once per-core LX usage is high enough, empirically ~1600KB), the backend's allocation overwrites it mid-op.

We have not confirmed whether this address is fixed regardless of op/shape or happens to coincide across the cases tested.

### What did you expect to happen?

If additional buffer is needed beyond the partition boundary, a compilation error or alarm should be raised.

### How can we reproduce it?

```python
import torch, torch_spyre

def fn(x):
    return torch.mean(x * x, dim=-1) + 1e-8

torch.manual_seed(0xAFFE)
x_cpu = torch.randn(32, 512, HIDDEN, dtype=torch.float16)  # HIDDEN = 1536 or 1600
cpu_result = fn(x_cpu)
result = torch.compile(fn)(x_cpu.to(device="spyre")).to("cpu")
torch.testing.assert_close(result, cpu_result, atol=1e-3, rtol=1e-2)
```
**a) With `+ 1e-8`:** both `HIDDEN=1536` and `HIDDEN=1600` **fail** (10.4% and 6.2% mismatch respectively). Here `mean`'s output is consumed by `+eps` in the same fused kernel, so it also gets LX-pinned — pinned buffers: `mul` output (1536KB or 1600KB) + `mean` output (64KB), concurrently live, peak 1600KB or 1664KB either way ≥ the failing threshold.

**b) Without `+ 1e-8`** (`return torch.mean(x * x, dim=-1)`): `HIDDEN=1536` **passes**, `HIDDEN=1600` **fails**. With no downstream consumer, `mean`'s output is single-use and stays in HBM — only `mul`'s output is LX-pinned. So peak LX usage equals `mul`'s output size alone: 1536KB (passes) or 1600KB (fails). This isolates the bug to `mul`'s buffer alone reaching 1600KB, with no `mean`-LX involvement at all.

Also reproduces with a pure pointwise chain (no reduction): `y = -x; z =
y + y + 1e-8` at `hidden=1600`.

**How LX usage and the backend address were confirmed**

- Per-core LX pinning/peak usage: run with `SPYRE_INDUCTOR_LOG=1 SPYRE_INDUCTOR_LOG_LEVEL=DEBUG`, grep stderr for
  `scratchpad` — the `spyre.inductor.scratchpad.plan_solver` and  `.allocator` lines report exactly which buffers are LX-pinned and their  per-timestep resident size.
- Backend allocation addresses: run with `DXP_VERBOSE=1`. This produces `debug/sdsc_N/sdsc_N.out.out.out.json` under each op's directory in `<TORCHINDUCTOR_CACHE_DIR>/inductor-spyre/sdsc_*/debug/` — the `scheduleTree_` list in this file has an `allocate` node per buffer with a concrete `startAddressCoreCorelet_` address. The colliding address (`1625344`) was found this way in both the `mean` case and an unrelated pointwise `neg` case.

A full `/tmp/torchinductor_cliu/` capture from one failing run (`HIDDEN=1600`, with `+ 1e-8`, compiled with `DXP_VERBOSE=1`) is attached.  The relevant file is `inductor-spyre/sdsc_fused_add_mean_mul_0_*/debug/sdsc_1/sdsc_1.out.out.out.json` (the `mean` op), where `allocate_lds1_lx` sits at address `1625344`, inside `mul`'s still-live output buffer spanning `[0, 1600KB)`.

### Any environmental details we need to know?

pod/env was rebuilt on Jul13.

### Anything else we need to know?

A temporary guard is added in PR #3154, constraining the frontend max usage to 80%*2MB - 100KB. 

### Relevant log output

```Shell

```

### PR DESCRIPTION
**Pull request by AdnanHoque**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Torch and DXP must agree on the exact LX ownership boundary. Torch currently estimates its available capacity using a hard-coded guard and may plan with a different fraction than the one passed to DXP. This can waste usable LX or cause the two allocators to disagree.
This PR calculates Torch’s budget from DXP’s actual usable LX capacity, applies the shared alignment rules, validates the configured fraction, and forwards that same value to DXP.

Fixes #3222

#### Which issue(s) this PR is related to:

https://github.com/torch-spyre/torch-spyre/issues/2787
https://github.com/torch-spyre/torch-spyre/issues/3049 
https://github.com/torch-spyre/torch-spyre/issues/3222
#3216 
#### Special notes for your reviewer:

This is part of a broader effort to size https://github.com/torch-spyre/torch-spyre/pull/2939 into several smaller PRs.

### CODE REVIEW

File: torch_spyre/execution/async_compile.py [hunk 1]
```diff
@@ -78,6 +79,15 @@ def sdsc(
 
         # Invoke backend compiler of SDSC Bundle
         with torch.profiler.record_function(f"dxp_standalone:{kernel_name}"):
-            subprocess.run(["dxp_standalone", "-d", output_dir], check=True)
+            backend_env = os.environ.copy()
```
Comment (chichun-charlie-liu) [inline torch_spyre/execution/async_compile.py]: isn't this `backend_env` we copied and fed to `dxp_standalone` here the same as the one `dxp_standalone` will see a few lines later? if we are worried about that some env var had changed, this copy should happen earlier?

**Review by cyang49 — COMMENTED** [review]
There are 2 things in this PR:
1. There's an alignment mismatch that could be the root cause of the issue found in #3222 @chichun-charlie-liu could you verify if this resolves the issue?
2. This PR changes how dxp_standalone is called by explicitly propagate DXP_LX_FRAC_AVAIL. It is unclear whether this is needed, as the env should be inherited by subprocesses. Please provide further justification.

File: torch_spyre/_inductor/config.py [hunk 2]
```diff
@@ -29,6 +29,9 @@
 
 allow_all_ops_in_lx_planning: bool = False
 
+# Fraction of Deeptools' post-program/debug LX capacity reserved for the DXP
+# backend. The Torch frontend owns ``1 - dxp_lx_frac_avail`` from address zero.
+# Torch forwards the configured value to DXP so their reservations cannot drift.
```
Comment (cyang49) [inline torch_spyre/_inductor/config.py]: no need to modify this file just to add the comment. it's already sufficiently explained in allocator.py comments

File: torch_spyre/execution/async_compile.py (hunk 1 above)
Comment (cyang49) [reply torch_spyre/execution/async_compile.py]: if this is only for the added tests, consider doing this in the test code
```
with config.patch({"dxp_lx_frac_avail": 0.375}), patch.dict(
    os.environ, {"DXP_LX_FRAC_AVAIL": "0.375"}
):
    ...
```

File: torch_spyre/_inductor/scratchpad/allocator.py [hunk 3]
```diff
@@ -56,6 +56,7 @@
 from torch_spyre._inductor.scratchpad.firstfit_bestfit_solver import (
     BestFitLayoutSolver,
     FirstFitLayoutSolver,
+    round_up_to_alignment,
```
Comment (isuruf) [inline torch_spyre/_inductor/scratchpad/allocator.py]: Move this function to `torch_spyre._inductor.scratchpad.utils` and import from there?

File: torch_spyre/execution/async_compile.py (hunk 1 above)
Comment (AdnanHoque) [reply torch_spyre/execution/async_compile.py]: torch plans from config.dxp_lx_frac_avail, while DXP reads the environment, the configured value is fwd'd so both use the same LX boundary...

File: torch_spyre/_inductor/config.py (hunk 2 above)
Comment (AdnanHoque) [reply torch_spyre/_inductor/config.py]: removed

**Comment by AdnanHoque** [comment]
@cyang49 @chichun-charlie-liu 

confirmed this fixes #3222 at DXP_LX_FRAC_AVAIL=0.2. 

Forwarded the value because Torch config can differ from the inherited environment.

**Review by isuruf — APPROVED** [review]

File: torch_spyre/_inductor/scratchpad/allocator.py (hunk 3 above)
Comment (AdnanHoque) [reply torch_spyre/_inductor/scratchpad/allocator.py]: yes, agree! moved.

File: torch_spyre/_inductor/config.py (hunk 2 above)
Comment (AdnanHoque) [reply torch_spyre/_inductor/config.py]: removed

**Review by cyang49 — APPROVED** [review]
lgtm

**Review by dgrove-oss — APPROVED** [review]
PR #3266: fix(inductor): match DXP usable LX reservation

Summary

Replaces a hard-coded "100 KiB safety margin" heuristic in _lx_planning_size() with an exact reimplementation of Deeptools' own LX partitioning math (physical 2 MiB minus a 64 KiB program/debug reservation, split by dxp_lx_frac_avail, rounded up to the 128-byte allocation granularity), so Torch's frontend LX budget matches what DXP will actually enforce instead of guessing at a margin. Also validates the fraction is in [0, 1] and moves the shared round_up_to_alignment helper into scratchpad/utils.py.

Review

BLOCKERs

None found.

SUGGESTIONs

None — the change is small, self-contained, and well-tested.

QUESTIONs

None.

Verification notes

- Hand-verified the three test_matches_deeptools_frontend_reservation cases (fractions 0.0/0.2/1.0) arithmetically: _LX_TRACKER_CAPACITY_BYTES = 2_031_616, and round_up_to_alignment(int(2_031_616 * (1 - frac)), 128) produces exactly 2_031_616, 1_625_344, and 0 — all match the expected test values.
- Confirmed not 0.0 <= nan <= 1.0 evaluates to True in Python (chained comparisons with NaN are always False), so the nan case in test_rejects_invalid_backend_fraction correctly raises.
- Confirmed config.patch({...}) is an established pattern already used elsewhere in the test suite (test_lx_inplace_layout.py, test_log_passes.py, etc.), so the new tests' usage is consistent with project conventions.
- Traced all three call sites of _lx_planning_size() (allocator.py lines 1290, 2025, 2076 on current upstream/main) — none need changes since the function's signature and semantics (bytes available to the layout solver) are preserved.
- Checked all importers of firstfit_bestfit_solver (allocator.py, simulated_annealing.py, and three test files) — none import round_up_to_alignment from that module specifically, so relocating it to utils.py doesn't break anything.
- DCO: single commit, signed off by Adnan Hoque. No added lines exceed 88 chars.
- The PR description's claim of "forwards that same value to DXP" refers to the shared DXP_LX_FRAC_AVAIL env var that both Torch and the separate DXP compiler binary read independently — not code in this repo, so no additional forwarding logic was expected or missing here.
- CI: linters, DCO, Build torch-spyre wheel, and Check OOT Configs all pass; the full test matrix (including Test Scratchpad Patterns/Test Scratchpad Solver, the two files touched) was still running/pending at check time with no failures observed. Already has 2 approvals (cyang49, isuruf) and auto-merge enabled.

Verdict

APPROVE

**Comment by dgrove-oss** [comment]
@AdnanHoque -- please fix the signatures on the commit and we can reapprove and merge.

**Comment by AdnanHoque** [comment]
fixed
